File: kirun-java/src/main/java/com/fincity/nocode/kirun/engine/json/schema/convertor/NumberConvertor.java
package com.fincity.nocode.kirun.engine.json.schema.convertor;

import static com.fincity.nocode.kirun.engine.util.json.ConvertorUtil.handleUnConvertibleValue;

import java.util.List;

import com.fincity.nocode.kirun.engine.json.schema.Schema;
import com.fincity.nocode.kirun.engine.json.schema.convertor.enums.ConversionMode;
import com.fincity.nocode.kirun.engine.json.schema.type.SchemaType;
import com.google.gson.JsonElement;
import com.google.gson.JsonNull;
import com.google.gson.JsonPrimitive;

public class NumberConvertor {

	private static final String NOT_A = " is not a ";

	public static JsonElement convert(List<Schema> parents, SchemaType type, Schema schema, ConversionMode mode,
			JsonElement element) {

		if (element == null || element.isJsonNull())
			return handleUnConvertibleValue(parents, mode, element, getDefaultValue(schema),
					"Expected a Number but found null");

		if (!element.isJsonPrimitive()) {
			return handleUnConvertibleValue(parents, mode, element, getDefaultValue(schema),
					element + NOT_A + type.getPrintableName());
		}

		JsonPrimitive jp = element.getAsJsonPrimitive();

		if (jp.isString())
			try {
				jp = new JsonPrimitive(Double.parseDouble(jp.getAsString()));
			} catch (NumberFormatException exception) {
				return handleUnConvertibleValue(parents, mode, element, getDefaultValue(schema),
						element + NOT_A + type.getPrintableName());
			}

		Number number = extractNumber(type, jp, mode);

		if (number == null) {
			return handleUnConvertibleValue(parents, mode, element, getDefaultValue(schema),
					element + NOT_A + type.getPrintableName());
		}

		return new JsonPrimitive(number);
	}

	private static Number extractNumber(SchemaType schemaType, JsonPrimitive jp, ConversionMode mode) {

		if (!jp.isNumber())
			return null;

		try {
			return switch (schemaType) {
				case INTEGER -> isInteger(jp, mode) ? jp.getAsInt() : null;
				case LONG -> isLong(jp, mode) ? jp.getAsLong() : null;
				case DOUBLE -> jp.getAsDouble();
				case FLOAT -> isFloat(jp, mode) ? jp.getAsFloat() : null;
				default -> null;
			};
		} catch (NumberFormatException numberFormatException) {
			return null;
		}
	}

	private static boolean isInteger(JsonPrimitive jp, ConversionMode mode) {
		if (mode != ConversionMode.STRICT) {
			return jp.isNumber();
		}

		return jp.isNumber() && jp.getAsDouble() % 1 == 0;
	}

	private static boolean isLong(JsonPrimitive jp, ConversionMode mode) {
		if (mode != ConversionMode.STRICT) {
			return jp.isNumber();
		}

		double value = jp.getAsDouble();

		return value % 1 == 0 && value >= Long.MIN_VALUE && value <= Long.MAX_VALUE;
	}

	private static boolean isFloat(JsonPrimitive jp, ConversionMode mode) {
		if (mode != ConversionMode.STRICT) {
			return jp.isNumber();
		}

		double value = jp.getAsDouble();

		return value >= -Float.MAX_VALUE && value <= Float.MAX_VALUE;
	}

	private static JsonElement getDefaultValue(Schema schema) {
		try {
			Number value = schema.getDefaultValue().getAsNumber();
			return new JsonPrimitive(value);
		} catch (Exception e) {
			return JsonNull.INSTANCE;
		}
	}

	private NumberConvertor() {
	}
}

File: kirun-js/src/engine/json/schema/convertor/NumberConvertor.ts
import { Schema } from '../Schema';
import { SchemaType } from '../type/SchemaType';
import { ConversionMode } from './enums/ConversionMode';
import { isNullValue } from '../../../util/NullCheck';
import { ConvertorUtil } from '../../../util/json/ConvertorUtil';

export class NumberConvertor {
    public static convert(
        parents: Schema[],
        type: SchemaType,
        schema: Schema,
        mode: ConversionMode,
        element: any,
    ): number | null {
        if (isNullValue(element)) {
            return ConvertorUtil.handleUnConvertibleValueWithDefault(
                parents,
                mode,
                element,
                this.getDefault(schema),
                'Expected a number but found null',
            );
        }

        if (typeof element === 'object' || typeof element === 'boolean' || Array.isArray(element)) {
            return ConvertorUtil.handleUnConvertibleValueWithDefault(
                parents,
                mode,
                element,
                this.getDefault(schema),
                element + ' is not a ' + type,
            );
        }

        if (typeof element === 'string') {
            element = Number(element);
            if (isNaN(element)) {
                return ConvertorUtil.handleUnConvertibleValueWithDefault(
                    parents,
                    mode,
                    element,
                    this.getDefault(schema),
                    element + ' is not a ' + type,
                );
            }
        }

        const number: number | null = this.extractNumber(type, element, mode);

        if (number === null) {
            return ConvertorUtil.handleUnConvertibleValueWithDefault(
                parents,
                mode,
                element,
                this.getDefault(schema),
                element + ' is not a ' + type,
            );
        }

        return number;
    }

    private static extractNumber(
        schemaType: SchemaType,
        element: number,
        mode: ConversionMode,
    ): number | null {
        if (typeof element !== 'number') {
            return null;
        }

        switch (schemaType) {
            case SchemaType.INTEGER:
                return this.isInteger(element, mode) ? Math.floor(element) : null;
            case SchemaType.LONG:
                return this.isLong(element, mode) ? Math.floor(element) : null;
            case SchemaType.DOUBLE:
                return element;
            case SchemaType.FLOAT:
                return this.isFloat(element, mode) ? element : null;
            default:
                return null;
        }
    }

    private static isInteger(value: number, mode: ConversionMode): boolean {
        if (mode !== ConversionMode.STRICT) {
            return typeof value === 'number';
        }
        return Number.isInteger(value);
    }

    private static isLong(value: number, mode: ConversionMode): boolean {
        if (mode !== ConversionMode.STRICT) {
            return typeof value === 'number';
        }
        return (
            Number.isInteger(value) &&
            value >= Number.MIN_SAFE_INTEGER &&
            value <= Number.MAX_SAFE_INTEGER
        );
    }

    private static isFloat(value: number, mode: ConversionMode): boolean {
        if (mode !== ConversionMode.STRICT) {
            return typeof value === 'number';
        }
        return value >= -Number.MAX_VALUE && value <= Number.MAX_VALUE;
    }

    private static getDefault(schema: Schema): number | null {
        if (typeof schema.getDefaultValue() === 'number') {
            return Number(schema.getDefaultValue);
        }
        return null;
    }
}
